### PR TITLE
pg_rxtx

### PR DESCRIPTION
**Pull request by outscale-jju**

### CODE REVIEW

File: src/rxtx.c [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	uint64_t bit;
+	uint16_t i;
+	struct rte_mbuf *tmp;
+	struct pg_rxtx_packet *b = state->rx_burst;
+	uint16_t cnt = 0;
+
+	/* prepare RX packets */
+	it_mask = pkts_mask;
+	for (; it_mask;) {
+		pg_low_bit_iterate_full(it_mask, bit, i);
```
Comment (outscale-mgo) [inline src/rxtx.c]: s/pg_low_bit_iterate/pg_low_bit_iterate_full/

File: src/rxtx.c [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+		      uint16_t edge_index, struct rte_mbuf **pkts,
+		      uint64_t pkts_mask, struct pg_error **error)
+{
+	struct pg_rxtx_state *state =
+		pg_brick_get_state(brick, struct pg_rxtx_state);
+
+	if (!state->rx)
+		return 0;
+	uint64_t it_mask;
+	uint64_t bit;
```
Comment (outscale-mgo) [inline src/rxtx.c]: this is useless

File: src/rxtx.c [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	uint16_t i;
+	struct rte_mbuf *tmp;
+	struct pg_rxtx_packet *b = state->rx_burst;
```
Comment (outscale-mgo) [inline src/rxtx.c]: s/b/rx_burst/

File: src/rxtx.c [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+		pg_low_bit_iterate_full(it_mask, bit, i);
+		tmp = pkts[i];
+		b[cnt].data = rte_pktmbuf_mtod(tmp, uint8_t *);
+		b[cnt].len = rte_pktmbuf_data_len(tmp);
+		cnt++;
+	}
+
+	/* pass burst to user */
+	state->rx(brick,
+		  state->rx_burst,
```
Comment (outscale-mgo) [inline src/rxtx.c]: s/state->rx_burst/rx_burst/

File: src/rxtx.c [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+		     struct pg_error **error)
+{
+	struct pg_rxtx_state *state =
+		pg_brick_get_state(brick, struct pg_rxtx_state);
+
+	if (!state->tx) {
+		*pkts_cnt = 0;
+		return 0;
+	}
+	struct pg_rxtx_packet *b = state->tx_burst;
```
Comment (outscale-mgo) [inline src/rxtx.c]: s/b/tx_burst/

File: src/rxtx.c [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+/* TX */
+static int rxtx_poll(struct pg_brick *brick, uint16_t *pkts_cnt,
+		     struct pg_error **error)
+{
+	struct pg_rxtx_state *state =
+		pg_brick_get_state(brick, struct pg_rxtx_state);
+
+	if (!state->tx) {
```
Comment (outscale-mgo) [inline src/rxtx.c]: please, don't mix code with variable declaration.

File: src/rxtx.c [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	struct pg_rxtx_packet *b = state->tx_burst;
+	struct rte_mbuf **mbuf = state->tx_mbuf;
+	struct pg_brick_side *s = &brick->side;
+	uint16_t count = 0;
+
+	/* let user write in packets */
+	state->tx(brick, b, &count, state->private_data);
+
+	/* write back packets lenght in mbuf */
+	for (int i = 0; i < count; i++)
```
Comment (outscale-mgo) [inline src/rxtx.c]: What is len is longer than the mbuf_size ?

File: include/packetgraph/rxtx.h [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef _PG_RXTX_H
+#define _PG_RXTX_H
+
+struct pg_rxtx_packet {
+	uint8_t *data;
+	uint32_t len;
```
Comment (outscale-mgo) [inline include/packetgraph/rxtx.h]: can't we use a pointer to the mbuf for the len ?

**Comment by outscale-jju** [comment]
fixed and merged

**Pull request closed without merging**